schedule() considers instruction map

MozammilQ opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
In issue #10837 , I discovered that the schedule of qiskit.compiler works fine for BackendV1 but doesn't work for BackendV2. See this [comment](https://github.com/Qiskit/qiskit/issues/10837#issuecomment-1750955255) for details.


### Details and comments
**Details**:
On inspection I discovered that measure function in macros.py requires backend, It uses backend to decide which of the two private methods _measure_v1 or _measure_v2 to call.
So, I decided to pass in this backend from the user to measure of macros.py
I did this as follows:
**1>>** I changed the definition of schedule_circuit in qiskit/qiskit/scheduler/schedule_circuit.py and added backend as Optional here.
**2>>** I used this backend and passed it to as_soon_as_possible and as_late_as_possible of qiskit/qiskit/scheduler/methods/basic.py, I also changed the definition of these two methods as_soon_as_possible and as_late_as_possible.
**3>>** I passed this backend to lower_gates by changing the definition of lower_gates in qiskit/qiskit/scheduler/lowering.py 
**4>>** I also changed the definition of the nested function get_measure_schedule of lowering.py.
**5>>**  Finally I passed in the backend to measure of macros.py

I have made backend in the definition of these changed methods Optional because I do not want the tests to fail, complaining the methods require backend as an additional argument.
I have reformatted the files that I changed and made sure that all the tests have passed! :) 


**Comments**:
qiskit/qiskit/compile/scheduler.py in the definition of schedule one cannot omit the backend because, how the inst_map is processed depends on it.Furthermore, there is no need to ask for inst_map and meas_map separately if you already have the backend with you!So, backend should not be Optional here! I think inst_map, meas_map and dt are redundant here, please correct me if I am wrong :)

Final result after issue fixed:
![fix_issue_10837](https://github.com/Qiskit/qiskit/assets/54737489/e0b96049-ad18-409c-bd37-4d20d345da6e)

MozammilQ committed:
fixed_issue_10837 by taking backend from the user and passing it to measure function of macros.py in qiskit/qiskit/pulse

MozammilQ commented:
sir, please see if my efforts a good enough :), please let me know if any additional changes are required :)

nkanazawa1989 commented:
@to24toro Could you please review the PR first? I'll merge if this looks good.

to24toro commented on qiskit/scheduler/lowering.py:
```diff
@@ -65,7 +70,9 @@ def lower_gates(circuit: QuantumCircuit, schedule_config: ScheduleConfig) -> Lis
 
-    def get_measure_schedule(qubit_mem_slots: Dict[int, int]) -> CircuitPulseDef:
+    def get_measure_schedule(
+        qubit_mem_slots: Dict[int, int], backend: Optional[Union[BackendV1, BackendV2]] = None
```
`get_measure_schedule` is a inner function and you can pass and use `backend`  without adding `backend`  as an argument.
```suggestion
    def get_measure_schedule(qubit_mem_slots: Dict[int, int]) -> CircuitPulseDef:
```

to24toro commented:
Thank you @MozammilQ .
It almost is good for me and will help solving the issue of scheduling with backendV2.

I have two additional requests for improvement.
- Add an argument of `backend` for the `sequence` in https://github.com/Qiskit/qiskit/blob/947e175edca77900cf1b0b266c55361bca0a7add/qiskit/scheduler/sequence.py#L28 and `lower_gates` in this `sequence`.
- Include explanations for added arguments to pass tests lint.

MozammilQ committed:
removing backend argumnet from inner function.

Since, get_measure_schedule is an innner function , so backend from lower_gates will be available to the inner function, so backend argument in the inner function is redundant :)

Co-authored-by: Kento Ueda <38037695+to24toro@users.noreply.github.com>

MozammilQ committed:
Fixed typo repetation of CircuitPulseDef, and removal of backend argument from the function call of get_measure_schedule

MozammilQ committed:
Adding relevant docstrings in the changed methods

MozammilQ commented:
> Thank you @to24toro for your suggestions.
> 
> I have done some improvements.
> 
>     * Add an argument of `backend` for the `sequence` in https://github.com/Qiskit/qiskit/blob/947e175edca77900cf1b0b266c55361bca0a7add/qiskit/scheduler/sequence.py#L28
**DONE**
>        and `lower_gates` in this `sequence`. 
**DONE**
> 
>     * Include explanations for added arguments to pass tests lint.
_I do not understand what I am supposed to do here, anyways I have added relevant docstring in the changed methods which I forgot to add, am I supposed to do anything else? adding explanations for added arguments in which file? and I have run all tests with tox and all tests have passed._

to24toro commented:
Thank you @MozammilQ . I will confirm it.
Anyway, you need to pass all the test to merge it (see :https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=50336&view=logs&jobId=86d13450-b8b4-5d9a-22f1-9f7e71062a71&j=86d13450-b8b4-5d9a-22f1-9f7e71062a71&t=a7f62dd9-9d78-5db3-3068-aed07887bd76). Could you please run `tox -eblack` and `tox -elint` in your PC and add the modification, referring to it?

MozammilQ committed:
fixed lint

MozammilQ commented:
@to24toro , :)
Aaah!! Now, I got it! thanks for the kind words, you explained it very well.
I ran `tox -eblack` and `tox -elint`, and did the required modifications in the last commit. Is it good this time? :) 
just for the information I have run tox -eblack/tox -elint on fedora 38.

to24toro commented:
Thank you @MozammilQ .
I think we are almost done. The left thing is to write test codes for backendV2 because the existing test codes related with your modification use backendV1. 

Especially, it's better to add the backendV2 version of https://github.com/Qiskit/qiskit/blob/65e8c1a7740be6bbd25c332da14df470f76d24d9/test/python/scheduler/test_basic_scheduler.py#L37 . These tests don't include handling the case that backend is backendV2.
However, this requires some knowledges and it might be tough.

to24toro commented on qiskit/scheduler/methods/basic.py:
```diff
@@ -37,6 +42,8 @@ def as_soon_as_possible(circuit: QuantumCircuit, schedule_config: ScheduleConfig
     Args:
         circuit: The quantum circuit to translate.
         schedule_config: Backend specific parameters used for building the Schedule.
+        backend: Pass in the backend used to build the Schedule, the backend could be BackendV1
```
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

to24toro commented on qiskit/scheduler/methods/basic.py:
```diff
@@ -87,6 +98,8 @@ def as_late_as_possible(circuit: QuantumCircuit, schedule_config: ScheduleConfig
     Args:
         circuit: The quantum circuit to translate.
         schedule_config: Backend specific parameters used for building the Schedule.
+        backend: Pass in the backend used to build the Schedule, the backend could be BackendV1
```
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

to24toro commented on qiskit/scheduler/schedule_circuit.py:
```diff
@@ -40,6 +44,8 @@ def schedule_circuit(
         circuit: The quantum circuit to translate.
         schedule_config: Backend specific parameters used for building the Schedule.
         method: The scheduling pass method to use.
+        backend: Pass in the backend used to build the Schedule, the backend could be BackendV1
```
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

to24toro commented on qiskit/scheduler/sequence.py:
```diff
@@ -35,14 +41,16 @@ def sequence(scheduled_circuit: QuantumCircuit, schedule_config: ScheduleConfig)
     Args:
         scheduled_circuit: The scheduled quantum circuit to translate.
         schedule_config: Backend specific parameters used for building the Schedule.
+        backend: Pass in the backend used to build the schedule, the backend could be either BackendV1
```
```suggestion
        backend: A backend used to build the Schedule, the backend could be BackendV1
```

MozammilQ commented:
@to24toro , please wait :)

TsafrirA commented:
@to24toro @MozammilQ 
No updates were made to any tests. Can't we add `FakeBackendV2` to the tests?

MozammilQ commented:
@TsafrirA , I have added many tests that uses BackendV2, there are a few tests that I am yet to add in, I am in talks with @to24toro , regarding that :)

MozammilQ committed:
Added tests to use BackendV2.

MozammilQ committed:
Apply suggestion in docstring.

Co-authored-by: Kento Ueda <38037695+to24toro@users.noreply.github.com>

MozammilQ committed:
Apply suggestion in docstring.

Co-authored-by: Kento Ueda <38037695+to24toro@users.noreply.github.com>

MozammilQ committed:
Apply suggestion in docstring.

Co-authored-by: Kento Ueda <38037695+to24toro@users.noreply.github.com>

MozammilQ committed:
Apply suggestion in docstring.

Co-authored-by: Kento Ueda <38037695+to24toro@users.noreply.github.com>

MozammilQ commented:
> Thank you @MozammilQ . I think we are almost done. The left thing is to write test codes for backendV2 because the existing test codes related with your modification use backendV1.
> 
> Especially, it's better to add the backendV2 version of
> 
> https://github.com/Qiskit/qiskit/blob/65e8c1a7740be6bbd25c332da14df470f76d24d9/test/python/scheduler/test_basic_scheduler.py#L37
> . These tests don't include handling the case that backend is backendV2.
> However, this requires some knowledges and it might be tough.

@to24toro , please see if this is ok :)

to24toro commented on test/python/scheduler/test_basic_scheduler.py:
```diff
@@ -494,3 +496,666 @@ def test_schedule_block_in_instmap(self):
+        backend = FakeBackend(None)
+        backend.defaults = backend.configuration
+        self.assertRaises(QiskitError, lambda: schedule(qc, backend))
+
```
```suggestion
```

This test is not related with the backend version and you can remove it.

to24toro commented on test/python/scheduler/test_basic_scheduler.py:
```diff
@@ -494,3 +496,666 @@ def test_schedule_block_in_instmap(self):
+        #       └────┘ ░ └────┘└────┘└───┘ ║ └╥┘
+        # c0: 2/═══════════════════════════╩══╩═
+        #                                  0  1
+
```
I don't understand that you change the number of gates from backendV1?
Isn't it enough to replace U2 gates with sx gates?

to24toro reviewed (changes requested):
Thank you @MozammilQ for adding tests. Your comments about the pulse durations is very clear and helps me understand it. I suggested the minor changes.

I would like you to add the release note about the bugfix (see: https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#release-notes)

MozammilQ committed:
added a release note

MozammilQ committed:
applied suggestions

MozammilQ committed:
checked lint

to24toro commented on releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml:
```diff
@@ -0,0 +1,32 @@
+    In test file:: test/python/scheduler/test_basic_scheduler.py
+      All the test methods that used BackendV1 has been replicated to use BackendV2.
+
+
```
```suggestion
```
Thank you for adding the release notes and sorry for my poor explanation. It would be sufficient to just add an item for "fixes". Also, please describe what was fixed in the fixes section. It would be good to briefly mention what was fixed, such as enabling scheduling using backendV2.

MozammilQ committed:
added suggestions in the release note

MozammilQ commented:
@to24toro , is everything good now? :)

to24toro reviewed (approved):
Thank you @MozammilQ . This Bugfix definitely helps users.

TsafrirA commented on qiskit/scheduler/lowering.py:
```diff
@@ -28,14 +28,19 @@
+def lower_gates(
+    circuit: QuantumCircuit,
+    schedule_config: ScheduleConfig,
+    backend: Optional[Union[BackendV1, BackendV2]] = None,
```
It's a little nitpicky, but all these unions could be replaced with `Backend`.

nkanazawa1989 commented on releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml:
```diff
@@ -0,0 +1,20 @@
+    `#10837 <https://github.com/Qiskit/qiskit/issues/10837>` for more information.
+
+
+
```
```suggestion
features:
  - |
    Enables circuit to pulse scheduling using BackendV2.
    
    .. code-block:: python
    
      # import a fake backend which is a sub-class of BackendV2.
      from qiskit.providers.fake_provider import FakePerth
      from qiskit.compiler.scheduler import schedule
      from qiskit.circuit import QuantumCircuit
      
      qc = QuantumCircuit(1, 1)
      qc.x(0)
      qc.measure(0,0)
      sched = schedule(circuits=qc, backend=FakePerth())

    Since, BackendV2 was not supported by the :func:`~qiskit.compiler.schedule` function, this caused
    schedule method to error out when the :code:`backend` argument is supplied with :class:`.BackendV2`. 
    Refer to `#10837 <https://github.com/Qiskit/qiskit/issues/10837>` for more information.

```

Cleanup. I think this should be classified into new feature, because the PR enables new object input which has not been supported previously.

nkanazawa1989 reviewed (dismissed):
Thanks @MozammilQ for contribution and @to24toro for review. I just added suggestion of the release note cleanup. You can push commit button if you like it. I'll approve the PR as is so that we can safely add this PR to merge queue before the release date.

MozammilQ committed:
Update releasenotes/notes/Adding-BackendV2-support-to-schedule-of-qiskit.compiler.scheduler-a0366d3397c73af0.yaml

Co-authored-by: Naoki Kanazawa <nkanazawa1989@gmail.com>

MozammilQ commented:
@nkanazawa1989 , does this look good now? :)

nkanazawa1989 committed:
Merge branch 'main' into fix_issue_10837

nkanazawa1989 reviewed (approved):
Thanks @MozammilQ this looks good to me. I'll add this to merge queue.
